pallets/phala/src/registry.rs

```diff
@@ -999,6 +1004,32 @@ pub mod pallet {
 			Ok(())
 		}
 
+		pub fn on_gk_launch_message_received(
+			message: DecodedMessage<GatekeeperLaunch>,
+		) -> DispatchResult {
+			if !message.sender.is_gatekeeper() {
+				return Err(Error::<T>::InvalidSender.into());
+			}
+
+			if let GatekeeperLaunch::MasterPubkeyOnChain(_) = message.payload {
+				let block_number = frame_system::Pallet::<T>::block_number();
+				let now = T::UnixTime::now().as_secs().saturated_into::<u64>();
+				GatekeeperLaunchedAt::<T>::put((block_number, now));
+				Self::deposit_event(Event::<T>::GatekeeperLaunched);
+			}
+			Ok(())
+		}
+
+		pub fn is_worker_registered_after_gk_launched(worker: &WorkerPublicKey) -> bool {
+			let Some(worker) = Workers::<T>::get(worker) else {
+				return false;
+			};
+			let Some((_, gk_launched_at)) = GatekeeperLaunchedAt::<T>::get() else {
+				return false;
+			};
+			worker.last_updated > gk_launched_at
```

Comment:
When `last_updated > gk_launched_at`, does it mean the worker must be aware of the gatekeeper? Is it possible that `last_updated` was updated but no notify is sent to GK?

Reply:
No notification is sent to GK. We just record the GK launched time on-chain and compare the workers' latest reg time against GK launched time when trying to add it to a pool.

Reply:
I was trying to double check `last_updated` can accurately reflect its registration date. With a re-register, I guess `last_updated` will be updated. In such case, will the updated worker still cause trouble when start mining?

Reply:
re-registering is OK. GK creates the worker's state machine on it first receives a registering event. re-registering will send the event too.

Reply:
Nice. Then it should work